Modified topology.proto to support for Python topology

taishi8117 opened the pull request:
Modified topology.proto so a component object can be defined in java or python. 
It is backward-compatible and easily extensible to add a support for another language.

This change is necessary for #1138

maosongfu commented on heron/proto/topology.proto:
```diff
@@ -71,9 +71,10 @@ message Config {
   // The name of the component. Like Tail-FlatMap
   required string name = 1;
-  // Only one of the below two are set
+  // Only one of the below three are set
```
Could we change "jave_class_name" and "jave_object" to a more generic name, rather than adding a new filed "python_class_name", which is not extensible.

billonahill commented on heron/proto/topology.proto:
```diff
@@ -71,9 +71,10 @@ message Config {
+  // Only one of the below three are set
   optional string java_class_name = 2;
   optional bytes java_object = 4;
+  optional string python_class_name = 5;
```
Could we use a Union type here with the enum discriminator? That would clarify which one is set.
https://developers.google.com/protocol-buffers/docs/techniques#union

maosongfu replied:
+1

billonahill replied:
Or what I just saw that @maosongfu suggested, but either way we should have decent typing for what kind of object we're specifying via an enum or other means.

taishi8117 replied:
@maosongfu @billonahill Ok, will do that.

kramasamy commented:
In addition to this - we might need additional fields to indicate - 

- Indicate whether the topology is written in C++/Java/Python/Go etc
- Higher level framework used as Summingbird, SQL, etc

This will be helpful later I believe.

billonahill commented:
@kramasamy I suggest we limit this change to what's needed to manage the python/java processes for now and follow up with the other higher-level info after clarifying our use cases around that more.

kramasamy commented:
@billonahill - I was more worried about the migration within Twitter.  If we can consolidate the changes into one release - it makes it little less painful.

taishi8117 commented:
@kramasamy @billonahill I added an enum to specify language, in which I added Java and Python for now, but it's easily extensible by adding new languages to the enum later.

billonahill commented:
I was advocating not blocking/coupling this PR with the feature you're talking about. It's best to separate features into different PRs/issue for better tracking.

We can put other things in this release if we want to, but let's deal with it separately.

billonahill commented on heron/proto/topology.proto:
```diff
@@ -71,9 +76,10 @@ message Config {
-  optional string java_class_name = 2;
-  optional bytes java_object = 4;
+  required Language lang = 5;
+  // Only one of the below three are set
```
Just paired on this. Will change Language to ComponentObjectSpec.

kramasamy commented:
Fine with me.

> On Jul 25, 2016, at 5:39 PM, Bill Graham <notifications@github.com> wrote:
> 
> I was advocating not blocking/coupling this PR with the feature you're talking about. It's best to separate features into different PRs/issue for better tracking.
> 
> We can put other things in this release if we want to, but let's deal with it separately.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

billonahill commented on heron/proto/topology.proto:
```diff
@@ -72,9 +78,11 @@ message Component {
   // The name of the component. Like Tail-FlatMap
   required string name = 1;
   // Only one of the below two are set
```
let's clarify this comment to describe how what's set below, and it's meaning can be determined by the value of spec.

billonahill commented:
👍

kramasamy commented:
@taishi8117 - can you look at the tracker code as well to ensure these changes do not affect? This is because tracker code use these protos as well.

maosongfu commented:
looks good to me. Please make sure the change does not break any existing components.

kramasamy commented:
👍

objmagic commented:
👍

nlu90 commented on heron/proto/topology.proto:
```diff
@@ -71,10 +77,16 @@ message Config {
+  optional bytes serialized_object = 4;
+
   required Config config = 3;
+  required ComponentObjectSpec spec = 5;
```
Adding required field to an existing proto file should always be prohibited. Hope we can find the problem before it's merged in the future.

taishi8117 replied:
@nlu90 Sorry for this, but #1180 fixed this problem.

nlu90 replied:
@taishi8117 thanks for the fix.
